src/AutoRest.CSharp/Mgmt/Output/MgmtObjectType.cs

```diff
@@ -2,129 +2,88 @@
-            CSharpType? inheritedType = null;
 
-            var sourceBaseType = ExistingType?.BaseType;
-            if (sourceBaseType != null &&
-                sourceBaseType.SpecialType != SpecialType.System_ValueType &&
-                sourceBaseType.SpecialType != SpecialType.System_Object &&
-                _typeFactory.TryCreateType(sourceBaseType, out CSharpType? baseType))
+            foreach (var property in DefinedProperties)
             {
-                inheritedType = baseType;
+                if (!baseProperties.Contains(property.Declaration.Name))
+                    yield return property;
             }
-            else
-            {
-                var objectSchemas = OjectSchema.Parents!.Immediate.OfType<ObjectSchema>().ToArray();
-
-                ObjectSchema? selectedSchema = null;
-
-                foreach (var objectSchema in objectSchemas)
-                {
-                    // Take first schema or the one with discriminator
-                    selectedSchema ??= objectSchema;
-
-                    if (objectSchema.Discriminator != null)
-                    {
-                        selectedSchema = objectSchema;
-                        break;
-                    }
-                }
+        }
 
-                if (selectedSchema != null)
-                {
-                    CSharpType type = _typeFactory.CreateType(selectedSchema, false);
-                    Debug.Assert(!type.IsFrameworkType);
-                    inheritedType = type;
-                }
-            }
+        protected override CSharpType? CreateInheritedType()
+        {
+            CSharpType? inheritedType = base.CreateInheritedType();
 
             var typeToReplace = inheritedType?.Implementation as ObjectType;
             if (typeToReplace != null)
             {
-                var match = InheritanceChoser.GetExactMatch((MgmtObjectType)typeToReplace);
+                var match = InheritanceChoser.GetExactMatch((MgmtObjectType)typeToReplace, DefinedProperties);
                 if (match != null)
                 {
                     inheritedType = match;
                 }
             }
-            return inheritedType == null ? InheritanceChoser.GetSupersetMatch(this) : inheritedType;
+            return inheritedType == null ? InheritanceChoser.GetSupersetMatch(this, DefinedProperties) : inheritedType;
+        }
+
+        protected override IEnumerable<ObjectTypeConstructor> BuildConstructors()
+        {
+            if (!(Inherits is not null && Inherits.Implementation is SystemObjectType))
+                yield return InitializationConstructor;
+
+            if (!IncludeDeserializer)
+            {
+                yield break;
+            }
+
+            // Skip serialization ctor if they are the same
+            if (!InitializationConstructor.Parameters
+                .Select(p => p.Type)
+                .SequenceEqual(SerializationConstructor!.Parameters.Select(p => p.Type)))
```

Comment:
This used to be:
-                .SequenceEqual(SerializationConstructor.Parameters.Select(p => p.Type)))
+                .SequenceEqual(SerializationConstructor!.Parameters.Select(p => p.Type)))

Could SerializationConstructor be null here?

Reply:
Its null for the SystemObjectType path and not null in the SchemaObjectType path.  So in this class its safe that it won't be null.

Reply:
I really do not like this "it can be null in one case and not null in another", it is really easy to trip in a future PR. However it is in Mgmt code, so I'm not blocking here.

Reply:
It is no longer null based on some changes I made in another area